pkg/operator/ceph/nfs/config.go

```diff
@@ -126,3 +133,231 @@ RADOS_URLS {
+	lockName := AppName
+
+	// planning to perform 2 commands after the lock: get object, and optionally write object, so
+	// use lock timeout of 2x the normal ceph command timeout
+	cookie, err := cephclient.RadosLockObject(context, clusterInfo,
+		radosPool, radosNamespace, objectName, lockName, exec.CephCommandsTimeout*2)
+	if err != nil {
+		return err // already a good err message
+	}
+	defer func() {
+		err := cephclient.RadosUnlockObject(context, clusterInfo,
+			radosPool, radosNamespace, objectName, lockName, cookie)
+		if err != nil {
+			logger.Infof("failed to unlock rados object, but since the lock has a timeout, we will continue. %v", err)
+		}
+	}()
+
+	cmd := cephclient.NewRadosCommand(context, clusterInfo,
+		append(radosFlags, "get", objectName, tempFile.Name()))
+	if _, err := cmd.RunWithTimeout(exec.CephCommandsTimeout); err != nil {
+		return errors.Wrapf(err, "failed to get object %s", objInfoString)
+	}
+
+	rawObj, err := ioutil.ReadAll(tempFile)
+	if err != nil {
+		return errors.Wrapf(err, "failed to read object %s from temp file", objInfoString)
+	}
+
+	if strings.Contains(string(rawObj), configBlock) {
+		logger.Debugf("rados object %s already has config block: %s", objInfoString, configBlock)
+		return nil
+	}
+	logger.Debugf("rados object %s will have config block prepended: %s", objInfoString, configBlock)
+
+	newConfig := fmt.Sprintf("%s%s", configBlock, string(rawObj))
+	if err := ioutil.WriteFile(tempFile.Name(), []byte(newConfig), fs.FileMode(0644)); err != nil {
+		return errors.Wrapf(err, "failed to write new config content for object %s to temp file", objInfoString)
+	}
+
+	cmd = cephclient.NewRadosCommand(context, clusterInfo,
+		append(radosFlags, "put", objectName, tempFile.Name()))
+	if _, err := cmd.RunWithTimeout(exec.CephCommandsTimeout); err != nil {
+		return errors.Wrapf(err, "failed to set new config content on object %s", objInfoString)
+	}
+
+	return nil
+}
+
+// remove a config block from the rados object in an atomic function
+// uses rados locks to ensure the changes we make here aren't going to race with changes that might
+// be made by ceph mgr, ceph itself, or a user manually
+func atomicRemoveFromConfigObject(context *clusterd.Context, clusterInfo *cephclient.ClusterInfo,
+	radosPool, radosNamespace, objectName, configBlock string,
+) error {
+	tmpFilePattern := fmt.Sprintf("%s_%s_%s_remove", radosPool, radosNamespace, objectName)
+	objInfoString := fmt.Sprintf("rados://%s/%s/%s", radosPool, radosNamespace, objectName)
+
+	// read object into temp file
+	tempFile, err := ioutil.TempFile("/tmp", tmpFilePattern)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create temp file for %s", objInfoString)
+	}
+	defer tempFile.Close()
+
+	radosFlags := []string{
+		"--pool", radosPool,
+		"--namespace", radosNamespace,
+	}
+
+	// acquire lock to ensure no other processes (user, ceph, or rook) are racing each other
+	lockName := AppName
```

Comment:
It's just a single lock for any client in this rook cluster, right? What else besides the operator could be contending for the lock?

Reply:
Ceph-CSI can create/delete NFS-volumes, which the creates/deletes the config files in RADOS through Ceph Mgr. I guess it is possible the volume is deleted while the config is modified?